src/Spago/Config/AST.hs

```diff
@@ -0,0 +1,964 @@
+--     but we can make assumptions about it. See the `EncounteredEmbed` constructor for `UpdateResult`.
+-- - ListLit - ["a", "literal", "list", "of", "values"]
+--     This is what will often be the expression associated with the "dependencies" key.
+--     Updating it merely means adding the new packages to its list.
+-- - ListAppend - `["list1"] # ["list2"]`
+--     Similar to `ListLit` except it appends two list expressions together.
+--     If it doesn't contain a `ListLit` (e.g. `expr1 # expr2`), we can wrap the entire
+--     expression in a `ListAppend originalExpr listExprWithNewPkgs`.
+-- - RecordLit - `{ key = value }`
+--     This is ultimately what we're looking for, so we can update the `dependencies`' field's list.
+--     However, due to supporting `Field`, which selects values within a record,
+--     we need to support looking up other keys besides the "dependencies" key.
+-- - Field recordExpr selection - `{ key = { dependencies = ["bar"] } }.key`
+--     This can produce a record expression. Depending on the record expression,
+--     we might need to update the values within the record expression
+--     that are ultimately exposed via the key (e.g. `{ config = { ..., dependencies = []} }.config`)
+-- - Prefer recordExpr overrides - `{ dependencies = ["foo"] } // { dependencies = ["bar"] }`
+--     This produces a record expression. If we update the `recordExpr`
+--     arg and its `dependencies` is overridden by `overrides`, then the update is pointless.
+--     If the `overrides` value overrides something irrelevant to the record's `dependencies` field,
+--     then we need to update the `recordExpr`. So, we need to try to update the `overrides` arg first
+--     and only if that fails do we attempt to update the `recordExpr`.
+-- - With recordExpr field update - `{ dependencies = ["foo"] } with dependencies = ["bar"]`
+--     This produces a record expression. Similar to `Prefer`,
+--     we should attempt to update the `update` value first before attempting to update `recordExpr`.
+-- - Let binding expr - `let binding = value in expr`
+--     The expression we need to update will often be in the `expr` (i.e. after the 'in' keyword;
+--     ``let src = [ "src" ] in { ..., dependencies = [ "old" ], sources = src }`).
+--     However, we might need to update the expression associated with the bound variable name.
+--     (e.g. `let config = { dependencies = [ "foo" ] } in config`). We won't know
+--     until we have looked at what the "dependencies" key refers to before finding out that
+--     the binding for the variable name `config` is what we need to update.
+--
+-- Since this is modifying the raw AST and might produce an invalid configuration file,
+-- the returned expression should be verified to produce a valid configuration format.
+modifyRawAST' :: HasLogFunc env => Text -> AstUpdate -> Expr -> RIO env Expr
+modifyRawAST' initialKey astMod originalExpr = do
```

Comment:
Does this function need to be in `RIO`? It looks like all branches are wrapped in `pure`, so maybe we can make it pure?

Reply:
We can make it pure. If we kept it in `RIO`, then we could add a 'logDebug` to show what decisions were made in various places to debug possible errors if `spago install` fails.

Reply:
I haven't addressed this yet because we might want to keep this impure initially so that if there is a bug in my code, it's easier to debug as we can see the exact path taken.

Reply:
Yeah sounds good, let's keep it in RIO